src/FffGcodeWriter.cpp

```diff
@@ -1425,155 +1425,237 @@ bool FffGcodeWriter::processSkinAndPerimeterGaps(const SliceDataStorage& storage
+        infill_comp.generate(gap_polygons, gap_lines);
+        if (gap_lines.size() > 0)
+        {
+            added_something = true;
+            setExtruder_addPrime(storage, gcode_layer, extruder_nr);
+            gcode_layer.setIsInside(true); // going to print stuff inside print object
+            gcode_layer.addLinesByOptimizer(gap_lines, &mesh_config.perimeter_gap_config, SpaceFillType::Lines);
+        }
     }
+    return added_something;
+}
 
-    // add insets
-    const Polygons* inner_skin_outline = nullptr;
-    int offset_from_inner_skin_outline = 0;
-    if (pattern != EFillMethod::CONCENTRIC)
+void FffGcodeWriter::processSkinInsets(const SliceDataStorage& storage, LayerPlan& gcode_layer, const SliceMeshStorage& mesh, const int extruder_nr, const PathConfigStorage::MeshPathConfigs& mesh_config, const SkinPart& skin_part, bool& added_something) const
+{
+    const int wall_0_extruder_nr = mesh.getSettingAsExtruderNr("wall_0_extruder_nr");
+    // add skin walls aka skin perimeters
+    if (extruder_nr == wall_0_extruder_nr)
     {
-        // add skin walls aka skin perimeters
-        if (extruder_nr == wall_0_extruder_nr)
+        for (const Polygons& skin_perimeter : skin_part.insets)
         {
-            for (const Polygons& skin_perimeter : skin_part.insets)
+            if (skin_perimeter.size() > 0)
             {
-                if (skin_perimeter.size() > 0)
-                {
-                    added_something = true;
-                    setExtruder_addPrime(storage, gcode_layer, layer_nr, extruder_nr);
-                    gcode_layer.setIsInside(true); // going to print stuff inside print object
-                    gcode_layer.addPolygonsByOptimizer(skin_perimeter, &mesh_config.insetX_config); // add polygons to gcode in inward order
-                }
+                added_something = true;
+                setExtruder_addPrime(storage, gcode_layer, extruder_nr);
+                gcode_layer.setIsInside(true); // going to print stuff inside print object
+                gcode_layer.addPolygonsByOptimizer(skin_perimeter, &mesh_config.insetX_config); // add polygons to gcode in inward order
             }
         }
-
-        // determine inner_skin_outline
-        if (skin_part.insets.size() > 0)
-        {
-            inner_skin_outline = &skin_part.insets.back();
-            offset_from_inner_skin_outline = -mesh_config.insetX_config.getLineWidth() / 2;
-        }
     }
+}
 
-    if (inner_skin_outline == nullptr)
+void FffGcodeWriter::processSkinPartTopMostInfillGeneratePerimeterGaps(const SliceDataStorage& storage, LayerPlan& gcode_layer, const SliceMeshStorage& mesh, const int extruder_nr, const PathConfigStorage::MeshPathConfigs& mesh_config, const SkinPart& skin_part, int skin_angle, const bool generate_perimeter_gaps, Polygons& concentric_perimeter_gaps, bool& added_something) const
```

Comment:
What you are asking for is a major refactor.
I don't think it's a good idea to make this issue bigger than it already is.

I know this architecture is rather suboptimal, but we shouldn't want to fix all the architecture touched by this single issue. The issue might then explode because each time we fix some architecture, some new piece of code will be touched of which the architecture should be changed.